core_v14/src/decoder/mod.rs

```diff
@@ -254,50 +272,149 @@ impl Decoder {
-		if !data.is_empty() {
-			return Err(DecodeError::LateEof(data.len(), ext));
-		}
+	/// Decode the SCALE encoded data that you would get signed, which conceptually takes the form
+	/// `(call_data, signed_extensions, additional_signed)`.
+	pub fn decode_signer_payload<'a>(&'a self, data: &mut &[u8]) -> Result<SignerPayload<'a>, DecodeError> {
+		let call_data = self.decode_call_data(data)?;
+		let signed_extensions = self.decode_signed_extensions(data)?;
+		let additional_signed = self.decode_additional_signed(data)?;
+		let extensions = signed_extensions
+			.into_iter()
+			.zip(additional_signed)
+			.map(|((name, extension), (_, additional))| (name, SignedExtensionWithAdditional { additional, extension }))
+			.collect();
+
+		Ok(SignerPayload { call_data, extensions })
+	}
+
+	/// Decode a SCALE encoded extrinsic signature.
+	fn decode_signature<'a>(&'a self, data: &mut &[u8]) -> Result<ExtrinsicSignature<'a>, DecodeError> {
+		let address = <MultiAddress<AccountId32, u32>>::decode(data)?;
+		let signature = MultiSignature::decode(data)?;
+		let extensions = self.decode_signed_extensions(data)?;
+
+		Ok(ExtrinsicSignature { address, signature, extensions })
+	}
+
+	/// Decode the signed extensions.
+	fn decode_signed_extensions<'a>(&'a self, data: &mut &[u8]) -> Result<Vec<(&'a str, Value)>, DecodeError> {
+		self.metadata
+			.extrinsic()
+			.signed_extensions()
+			.iter()
+			.map(|ext| {
+				let val = decode_type_by_id(data, &ext.ty, self.metadata.types())?;
+				let name = &*ext.identifier;
+				Ok((name, val))
+			})
+			.collect()
+	}
+
+	/// Decode the additional signed data. This isn't used for decoding extrinsics, but instead for
+	/// decoding the data that you'd sign.
+	fn decode_additional_signed<'a>(&'a self, data: &mut &[u8]) -> Result<Vec<(&'a str, Value)>, DecodeError> {
+		self.metadata
+			.extrinsic()
+			.signed_extensions()
+			.iter()
+			.map(|ext| {
+				let val = decode_type_by_id(data, &ext.additional_signed, self.metadata.types())?;
+				let name = &*ext.identifier;
+				Ok((name, val))
+			})
+			.collect()
+	}
+}
+
+/// Decoded call data and associated type information.
+#[derive(Debug, Clone, PartialEq)]
+pub struct CallData<'a> {
+	/// The name of the pallet
+	pub pallet_name: &'a str,
+	/// The type information for this call (including the name
+	/// of the call and information about each argument)
+	pub ty: &'a scale_info::Variant<scale_info::form::PortableForm>,
+	/// The decoded argument data
+	pub arguments: Vec<Value>,
+}
+
+/// Decoded call data and associated type information.
+#[derive(Debug, Clone, PartialEq)]
+pub struct CallDataOwned {
+	/// The name of the pallet
+	pub pallet_name: String,
+	/// The type information for this call (including the name
+	/// of the call and information about each argument)
+	pub ty: scale_info::Variant<scale_info::form::PortableForm>,
+	/// The decoded argument data
+	pub arguments: Vec<Value>,
+}
```

Comment:
I thought about using `Cow`/`Borrow` in some way, but in order for that to work, you have to be able to `borrow` from the owned version to get the borrowed version, which I don't think is possible here.

I did wonder about some other kind of paramterising to avoid the duplication, but I didn't have any good ideas offhand that wouldn't just make the code more complex. Definitely open to ideas though if you can think up something that works and isn't too much of a complexity overhead :)

Reply:
@insipx do you have any opinions on whether the Owned structs in this PR are useful? I'm inclined to remove them and leave that to end users offhand.

Reply:
I think we can actually use Cows here and get the best of both worlds. I think we could impl them in this style:

Reply:
You were absolutely right; I converted to using Cow and everything worked; no more owned structs :)

Reply:
(Also, I had a go with beef, but it's specific to certain types where the tag and capacity can be merged to save space, so will stick with `std::borrow::Cow` :) )